pkg_resources/__init__.py

```diff
@@ -840,6 +847,7 @@ def resolve(self, requirements, env=None, installer=None,
             # Register the new requirements needed by req
             for new_requirement in new_requirements:
                 required_by[new_requirement].add(req.project_name)
+                extra_req_mapping[new_requirement] = req.extras
```

Comment:
This is buggy vs the sketch I put up in my fork - you're not unioning together all the extras. Consider A and B both depend on C - one on C[a] and one on C[b]. We want to activate C[a,b].